### ISSUE TITLE
Support Disk Encryption Set for Node OS Disks in Karpenter

### ISSUE THREAD

**Issue by iamvighnesh** [Azure/karpenter-provider-azure#268]
### Tell us about your request

The users are requesting support for Disk Encryption Sets for Node OS Disks on Karpenter-managed nodes. 

Here's the existing feature for [AKS managed nodes with Disk Encryption Set for Node OS Disks](https://learn.microsoft.com/en-us/azure/aks/azure-disk-customer-managed-keys).

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

The users do not have a way to provide a customer-managed key for encrypting node OS disk and would like the ability to supply customer-managed keys to use for encryption at rest for both the OS for Karpenter-managed nodes.

### Are you currently working around this issue?

No workaround is available for this right now.

### Additional Context

- Do we also need to consider [supporting Host Encryption for the nodes](https://learn.microsoft.com/en-us/azure/aks/enable-host-encryption)?
- Right now, Disk Encryption Set can only be enabled during cluster creation. We need to investigate how that would affect the karpenter-managed nodes.
- What happens if the AKS cluster does not have Disk Encryption Set enabled?

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by tallaxes** [comment]
This is currently not supported. The first step would be to enable disk encryption with managed keys by default, which is what AKS does. Next step would be support for BYOK.

**Comment by iamvighnesh** [comment]
@tallaxes Would you like me create a separate issue for Disk Encryption using platform managed keys?

I am happy to pick this up and work on both the features.

**Comment by tallaxes** [comment]
Good idea, and appreciate offer of help! Let me dig a little bit into the differences between this (server-side encryption) and host-based encryption, to see how to break this down best - and maybe set priorities.

**Comment by iamvighnesh** [comment]
Let me know if you would like me to split this into two: [Disk Encryption](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/disk-encryption-overview) and [Host Encryption](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-enable-host-based-encryption-portal?tabs=azure-powershell).

Happy to pick either.

**Comment by seizste** [comment]
Hello @tallaxes @Bryce-Soghigian ,

looks like @michaelerne supplied a pull request that implements the functionality to implement Disk Encryption as suggested by @iamvighnesh . Would be great if you could have a look at this, as not being able to use CMK currently blocks us to rollout Karpenter in our Environment.

**Comment by asychev** [comment]
@tallaxes Are there any ETA for this feature supported in NAP?

**Comment by wdarko1** [comment]
> Are there any ETA for this feature supported in NAP?

@asychev This feature is scheduled to be released in late January/early February 2026.

**Comment by ilmax** [comment]
@wdarko1 any way to track the rollout of this?
